kombu/asynchronous/hub.py

```diff
@@ -242,6 +245,12 @@ def _unregister(self, fd):
         except (AttributeError, KeyError, OSError):
             pass
 
+    def _pop_ready(self):
+        with self._ready_lock:
+            ready = self._ready
+            self._ready = set()
```

Comment:
Is this an optimization or a source of future problems?
A set is theoretically unordered but a list is ordered which is what we used before.

In any case, I think the main problem with this code is it's design.
ready should not be a set but a priority queue with unique items. See [implementation example](https://gist.github.com/Dobiasd/37705392b4aaa3a3539ba1a61efec6b6).
This data structure will allow us to re-order the execution of events later on. The bonus is that it is thread-safe and it doesn't pollute this class with more responsibilities it shouldn't have.

Reply:
> In any case, I think the main problem with this code is it's design.

I agree. The reason I decided to take the approach in this PR is that it is a small change. I can't implement a better approach due to a lack of experience in Kombu.

Reply:
Feel free to close this PR and implement it in a much better way. Or merge this PR, release a bugfix and prepare a better implementation later.

Reply:
I think I see your point of view better now. It should work. I'm going to merge this.

Some promises have to be called as soon as possible since their ETA related for example so they have to be prioritized over regular tasks. I guess it doesn't matter now and we can open a ticket about this later